docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -0,0 +1,110 @@
+> Yeah, AFAIK, 20 bytes should be collision resistance when the preimages are unique and not malleable. A space of 2^160 would expect some collision to be likely around 2^80 elements (birthday paradox). And if you want to find a collision for some existing element in the database, it is still 2^160. 2^80 only is if all these elements are written to state.
+
+> The good example you brought up was eg. a public key bytes being a valid public key on two algorithms supported by the codec. Meaning if either was broken, you would break accounts even if they were secured with the safer variant. This is only as the issue when no differentiating type info is present in the preimage (before hashing into an address).
+
+> I would like to hear an argument if the 20 bytes space is an actual issue for security, as I would be happy to increase my address sizes in weave. I just figured cosmos and ethereum and bitcoin all use 20 bytes, it should be good enough. And the arguments above which made me feel it was secure. But I have not done a deeper analysis.
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
+
+The current multisig public keys use amino serialization to generate the address. We will retain
+those public keys and their address formatting, and call them "legacy amino" multisig public keys
+in protobuf. We will also create multisig public keys without amino addresses to be described below.
+
+### Canonical Public Key Addresses
+
+Following on the discussion in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we propose the
+following approach.
+
+All public key types will have a unique protobuf message type such as:
+
+```proto
+package cosmos.crypto.sr25519;
+
+message PubKey {
+  bytes key = 1;
+}
+```
+ 
+All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
+These names are derived directly from .proto files in a standardized way and used
+in other places such as the type URL in `Any`s. Since there is an easy and obvious
+way to get this name for every protobuf type, we can use this message name as the
+key type prefix when creating addresses.
+
+We define the canonical address format for new (non-legacy) public keys as
+`Sha256(fmt.Sprintf("%s/%x, proto.MessageName(key), key.Bytes())[:20]`. This takes
+the first 20 bytes of an SHA-256 hash of a string with the proto message name for the key
+type joined by an `/` with the hex encoding of the key bytes.
```

Comment:
Actually, why bother with the `/` either? In #5694 we talked about `Sha256(KeyTypePrefix || Keybytes)[:20]` which is just pure concatenation. I guess because the prefix is a string not just a byte, so we need a domain separator? It seems like maybe we're optimizing for human readability of the pre-image; is that important for something?

Reply:
PubKeys are not meant to be human readable. They are long. That's why we are creating addresses.

So we don't even need to convert to string. We can concatenate bytes, without conversion to string.

Reply:
I've updated this ADR to define an `AddressHash` function which takes a string `prefix` and binary `contents` separating them by a null terminator `0` byte and which doesn't do hex-encoding of binary. Can you take a look again @ebuchman ?